compiler/noirc_evaluator/src/ssa/opt/brillig_entry_points.rs

```diff
@@ -0,0 +1,612 @@
+    ir::{
+        function::{Function, FunctionId},
+        instruction::Instruction,
+        value::Value,
+    },
+    Ssa,
+};
+
+use super::inlining::called_functions_vec;
+
+impl Ssa {
+    pub(crate) fn brillig_entry_point_analysis(mut self) -> Ssa {
+        if self.main().runtime().is_brillig() {
+            return self;
+        }
+
+        let brillig_entry_points = get_brillig_entry_points(&self.functions, self.main_id);
+        let entry_points = brillig_entry_points.keys().copied().collect::<HashSet<_>>();
+
+        let inner_call_to_entry_point = build_inner_call_to_entry_points(&brillig_entry_points);
+
+        let functions_to_clone_map =
+            build_functions_to_clone(&self.functions, inner_call_to_entry_point, entry_points);
+
+        let calls_to_update = build_calls_to_update(&mut self, functions_to_clone_map);
+
+        let mut new_functions_map = HashMap::default();
+        for (entry_point, inner_calls) in brillig_entry_points {
+            let new_entry_point =
+                new_functions_map.get(&entry_point).copied().unwrap_or(entry_point);
+
+            let function =
+                self.functions.get_mut(&new_entry_point).expect("ICE: Function does not exist");
+            update_function_calls(function, entry_point, &mut new_functions_map, &calls_to_update);
+
+            for inner_call in inner_calls {
+                let new_inner_call =
+                    new_functions_map.get(&inner_call).copied().unwrap_or(inner_call);
+
+                let function =
+                    self.functions.get_mut(&new_inner_call).expect("ICE: Function does not exist");
+                update_function_calls(
+                    function,
+                    entry_point,
+                    &mut new_functions_map,
+                    &calls_to_update,
+                );
+            }
+        }
+
+        self
+    }
+}
+
+fn build_functions_to_clone(
+    functions: &BTreeMap<FunctionId, Function>,
+    inner_call_to_entry_point: HashMap<FunctionId, Vec<FunctionId>>,
+    entry_points: HashSet<FunctionId>,
+) -> HashMap<FunctionId, Vec<(Function, FunctionId)>> {
+    let mut functions_to_clone_map: HashMap<FunctionId, Vec<(Function, FunctionId)>> =
+        HashMap::default();
+
+    let mut add_function_to_clone = |entry_point: FunctionId, inner_call: FunctionId| {
+        let cloned_function = Function::clone_no_id(&functions[&inner_call]);
+        functions_to_clone_map.entry(entry_point).or_default().push((cloned_function, inner_call));
+    };
+
+    for (inner_call, inner_call_entry_points) in inner_call_to_entry_point {
+        if inner_call_entry_points.len() > 1 {
+            for entry_point in inner_call_entry_points {
+                add_function_to_clone(entry_point, inner_call);
+            }
+        } else if entry_points.contains(&inner_call) {
+            add_function_to_clone(inner_call_entry_points[0], inner_call);
+        }
+    }
+
+    functions_to_clone_map
+}
```

Comment:
It's not entirely clear what the conditions to choose to clone a function are without reading the entire function. This could do with a renaming or documentation.